### ISSUE TITLE
mapTo documentation should be updated to explain "eager evaluation" behavior

### ISSUE THREAD

**Issue by vorong** [ReactiveX/rxjs#5526]
## Bug Report

In re this bug: https://github.com/ReactiveX/rxjs/issues/5525. When I ran into it, I checked the docs for both map (https://rxjs-dev.firebaseapp.com/api/operators/map) and mapTo (https://rxjs-dev.firebaseapp.com/api/operators/mapTo).

Doc for mapTo clearly says "Takes a constant value as argument, and emits that whenever the source Observable emits a value. In other words, ignores the actual source value, and simply uses the emission moment to know when to emit the given value."

But, it is not really using the emission moment to emit the value. It is emitting its value before the previous one has occurred. I understand the design decision to optimize for speed (since you are ignoring the value anyway), but at the very least, the documentation so be updated to be extremely clear about this point. Until someone has been bitten by this behavior once, there's no way for them to know the difference, and it's not immediately obvious that mapTo should have this optimization.

**Comment by kwonoj** [comment]
I don't see problem of existing documentation: it clearly says

1. it uses given constant `value`
2. it uses sources emission `moment`

2 and 1 is not conflicting - document never suggests it try to `evaluate` value at the moment of source emission. Instead it clarifies it uses given constant value but emit when source emits instead of emitting source's value.

**Comment by kwonoj** [comment]
>  It is emitting its value before the previous one has occurred

I'm not sure if I understood this correctly: when does `mapTo` actually emits value before previous one occurred? do you have an example?

**Comment by vorong** [comment]
The mapTo doc also says this: "Like map, but it maps every source value to the same output value every time." So, based on that, I would expect mapTo to behave like map.

Here's an example: https://stackblitz.com/edit/map-vs-mapto

**Comment by kwonoj** [comment]
I'd say `Like map, but it maps every source value to the same output value every time.` requires some clarification and agree this potentially able to be misread. I don't think other documentation states incorrect.

**Comment by cartant** [comment]
Closing this because I really don't see how the documentation is unclear or confusing.

**Comment by vorong** [comment]
@cartant @kwonoj  The documentation says: "Emits the given constant value on the output Observable every time the source Observable emits a value." But, if you take a look at my example (https://stackblitz.com/edit/map-vs-mapto), that is not entirely true: it also emits a constant value whenever it is tapped, even if the source observable has not yet emitted a value.

Per the explanation given in #5525, this is expected behavior, i.e. since mapTo doesn't actually need an input value, it eagerly evaluates and returns the constant value anyway. This is not indicated anywhere in the documentation.

**Comment by cartant** [comment]
Of course it's 'eagerly evaluated'. `mapTo` is a function, the argument passed to it is evaluated when when the function is called - i.e. when the observable is composed - not when the source emits. This argument-passing behaviour is fundamental and it's not the responsibility of this specific operator's documentation to explain how function calls behave.

**Comment by seanheinen** [comment]
For what it's worth I've spent hours on this exact _ambiguity_.

Yes functions calls are by nature "eager", but I think what @vorong is pointing out is that a `map` to a constant value is not simply replaceable with a `mapTo` operator, which in my interpretation the documentation suggests.

When i discovered these `mapTo` operators I figured: 
"Hey, anytime I'd ignore the source `map(() => ...)` I could just use the `mapTo` operator." Not true.

Having the constant being evaluated immediately is in hindsight obvious, however I'd argue that that very feature almost makes the use of these operators scarce at best.

e.g. This doesn't work (again, obvious in hindsight).

```
let foo = false;

of('test').pipe(
  tap(() => {
    foo = true;
  }),
  mapTo(foo),
  filter((next) => next)
).subscribe((next) => {
  console.log(next);
});
```

It's worth mentioning that the `_To` operators also require a completely different approach to unit testing, which has become enough of a headache that I've just avoided using them.

I'm a +1 for maybe pointing out this caveat somewhere.
